File: src/server/ServerWithConnectAndDisconnectCommand.java
package server;

import com.fasterxml.jackson.core.JsonProcessingException;
import network.client.udp.ConnectionStatus;
import network.client.udp.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.net.DatagramPacket;
import java.net.SocketException;

public abstract class ServerWithConnectAndDisconnectCommand extends ServerWithCommands {
    private static final Logger logger = LogManager.getLogger(ServerWithConnectAndDisconnectCommand.class);

    protected ServerWithConnectAndDisconnectCommand(int port, String commandFieldName) throws SocketException {
        super(port, commandFieldName);
    }

    protected abstract void handleConnectCommand(User user);

    protected abstract void handleDisconnectCommand(User user);

    protected ConnectionStatus handleConnectAndDisconnectCommands(DatagramPacket packet) throws JsonProcessingException {
        String command = getCommandFromJson(packet);
        User user = getUserFromPacket(packet);

        switch (command) {
            case "Connect":
                handleConnectCommand(user);
                logger.trace("User {} connected", user);
                return ConnectionStatus.JUST_CONNECTED;

            case "Disconnect":
                handleDisconnectCommand(user);
                logger.trace("User {} disconnected", user);
                return ConnectionStatus.DISCONNECTED;

            default:
                return ConnectionStatus.CONNECTED;
        }
    }
}

File: src/server/ServerWithConnectAndDisconnectCommand.kt
package server

import lib.net.udp.JsonHolder
import network.client.udp.ConnectionStatus
import network.client.udp.User
import org.apache.logging.log4j.kotlin.Logging

abstract class ServerWithConnectAndDisconnectCommand protected constructor(port: Int, commandFieldName: String) :
    ServerWithCommands(port, commandFieldName), Logging {

    protected abstract suspend fun handleWhenConnected(jsonHolder: JsonHolder)
    protected abstract fun handleConnectCommand(user: User)
    protected abstract fun handleDisconnectCommand(user: User)

    protected fun handleConnectAndDisconnectCommands(jsonHolder: JsonHolder): ConnectionStatus {
        val command = getCommandFromJson(jsonHolder)
        val user = jsonHolder.user

        return when (command) {
            "Connect" -> {
                handleConnectCommand(user)
                logger.trace("User $user connected")
                ConnectionStatus.JUST_CONNECTED
            }

            "Disconnect" -> {
                handleDisconnectCommand(user)
                logger.trace("User $user disconnected")
                ConnectionStatus.DISCONNECTED
            }

            else -> ConnectionStatus.CONNECTED
        }
    }

    override suspend fun handlePacket(jsonHolder: JsonHolder) {
        val status = handleConnectAndDisconnectCommands(jsonHolder)

        if (status == ConnectionStatus.CONNECTED) {
            handleWhenConnected(jsonHolder)
        }
    }
}
